Weighted civilian vehicle templates

jaj22 opened the pull request:
## What type of PR is this.
1. [X] Bug
2. [X] Enhancement

### What have you changed and why?
Since 2.2.2 we've been using automatic detection of civilian vehicles. This caused several issues, for example:
1. Some mods incorrectly categorized vehicles, eg. CUP carriers and RHS Ural trucks.
2. Each vehicle class has an equal chance of spawning, so specialised vehicles and those with multiple classnames for different paint jobs were too common, and basic vehicles with automated paint jobs were too rare.
3. Similarly, changing the enabled DLC or modset used could cause large variations in the types of vehicles available.
4. Even though 3CB has enough vehicles for a total conversion, the vanilla vehicles were still included, arguably diminishing the mod flavour.

This PR adds per-modset civilian templates as a fourth template type. Currently these are only used for specifying civilian vehicles, but they could be expanded to civilian units, weapon caches and vehicle item loadouts, for example. I reworked the template-loading code to what I thought was a far more readable format, but this is an optional change.

The civilian vehicles have been split into common, industrial, medical, repair and refuel categories. Each entry has a weight relative to the other vehicles in that category. In initVarServer, each category is filtered for DLC vehicles and then re-weighted so that for example, medical vehicles are 10x less common than trucks. These category weights could be set as parameters, but they're hardcoded in initvar for now.

Boats are also created with a weighted array, and filtered for DLC, but I didn't bother categorizing them as there aren't enough to bother. 

There's now a separate undercover vehicles list which is currently all civilian vehicles except the quadbike. Civilian quadbikes can now spawn, but don't allow undercover.

I created the 3CB list as a total conversion, with no vanilla ground vehicles included. Not really attached to the decision, it's just an example. You could instead mix in small quantities of vanilla vehicles, or mix them evenly.

### Please specify which Issue this PR Resolves.
closes #790, #727

### Is further testing or are further changes required?
1. [X] No
2. [ ] Yes (Please provide further detail below.)

General method needs approval though. Will purge some dead code in initVarServer if approved.

SonorousMeerkat commented:
Looks really good!
1 question; does a weight of 0 in the template mean never spawns or rarely spawns?

jaj22 commented:
@SonorousMeerkat 0 => never spawns, but will still get added to the undercover list. My concern with the Ikarus is that it's likely to block roads. Also doesn't make much sense having empty ones parked around: For future work, I considered having separate weighted lists for parked and moving vehicles, although the moving vehicle spawn system needs a major rework anyway.

Spoffy commented on A3-Antistasi/functions/Templates/fn_ifaModCompat.sqf:
```diff
@@ -28,5 +28,5 @@ lootWeapon deleteAt (lootWeapon find _x);
 } forEach _libStaticParts;
 
 //Add some more variety to the civilian vehicles.
-arrayCivVeh append ["LIB_Zis5v_w","LIB_Kfz1_Hood_sernyt"];
```
Please can you remove these comments? They'll just linger indefinitely if we leave them in now!

Spoffy commented on A3-Antistasi/functions/init/fn_initVarServer.sqf:
```diff
@@ -214,7 +214,7 @@ DECLARE_SERVER_VAR(arrayCivs, _arrayCivs);
 //      CIVILIAN VEHICLES       ///
 ////////////////////////////////////
 [2,"Creating vehicles list",_fileName] call A3A_fnc_log;
-
```
Please can you remove these comments? They'll just linger indefinitely if we leave them in now!

Spoffy commented on A3-Antistasi/functions/init/fn_initVarServer.sqf:
```diff
@@ -492,6 +493,44 @@ private _templateVariables = [
 	ONLY_DECLARE_SERVER_VAR_FROM_VARIABLE(_x);
 } forEach _templateVariables;
 
+call {
```
While this is definitely more readable, you've changed the logic of how we load RHS mods - It's no longer possible to just run USAF or AFRF - you need all 3 now, and I don't think that's something we should be doing.

jaj22 replied:
I had no idea that was an intentional or supported feature, but it looks like it was considered in the equipment filters, so fair enough.

There's a related question of how the faction-selection interface will work for the template rework, but I guess that can wait.

jaj22 commented:
Changes made:

- Reverted to old template selection method to re-enable vanilla/RHS mix & match
- Moved useful parts of old automated vehicle detection code into Auto_Civ template
- Removed a lot of dead code
